boot/bootstate20.go

```diff
@@ -453,24 +518,41 @@ func (ks20 *bootState20Kernel) revisions() (curSnap, trySnap snap.PlaceInfo, try
 
 func (ks20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
 	// call the generic method with this object to do most of the legwork
-	u, sn, err := selectSuccessfulBootSnap(ks20, update)
+	u20, sn, err := selectSuccessfulBootSnap(ks20, &ks20.bootState20Modeenv, update)
 	if err != nil {
 		return nil, err
 	}
 
-	// save this object inside the update to share bootenv / modeenv between
-	// multiple calls to markSuccessful for different snap types, but the same
-	// bootStateUpdate object
-	u.bootState20Kernel = *ks20
+	// XXX: this if arises because some unit tests rely on not setting up kernel
+	// details and just operating on the base snap but this situation would
+	// never happen in reality
+	if sn != nil {
+		// mark the kernel successful before rewriting the modeenv because if
```

Comment:
maybe all these comments that preceed things delayed to the commit should start with: On commit ...

Reply:
makes sense